File: sdk/rust/src/state.rs
//! Game state with perspective mapping and convenience methods.
//!
//! [`GameState`] owns a real engine `GameState` mirror. The host advances its
//! canonical state turn-by-turn; the SDK applies the same Advance deltas so
//! the two stay in sync. Verification is via `state_hash()` (Zobrist), which
//! the engine maintains incrementally on `process_turn`.
//!
//! Perspective mapping translates player1/player2 to my/opponent based on the
//! slot assigned by `HostMsg::Welcome` during the handshake.

use std::collections::HashMap;

use pyrat::{Coordinates, Direction, GameBuilder, MudMap};
use pyrat_engine_interface::pathfinding::FullPathResult;
use pyrat_engine_interface::Maze;
use pyrat_wire::Player;

use crate::GameSim;
use pyrat_protocol::{MatchConfig, TurnState};

/// SDK-facing game state. Built from `MatchConfig`, advanced each turn.
pub struct GameState {
    engine: pyrat::GameState,
    my_player: Player,

    // The protocol carries last_move per player; engine GameState doesn't track it.
    player1_last_move: Direction,
    player2_last_move: Direction,

    // Static config from MatchConfig (not part of the engine state).
    move_timeout_ms: u32,
    preprocessing_timeout_ms: u32,
}

impl GameState {
    /// Build from match configuration received during setup.
    ///
    /// `slot` is the player slot assigned by [`HostMsg::Welcome`]. Everything
    /// from `cfg.controlled_players` is ignored — the slot is authoritative.
    pub fn from_config(slot: Player, cfg: &MatchConfig) -> Result<Self, String> {
        let engine = build_engine(cfg)?;
        Ok(Self {
            engine,
            my_player: slot,
            player1_last_move: Direction::Stay,
            player2_last_move: Direction::Stay,
            move_timeout_ms: cfg.move_timeout_ms,
            preprocessing_timeout_ms: cfg.preprocessing_timeout_ms,
        })
    }

    /// Apply an Advance delta. Updates the engine via `process_turn`, which
    /// keeps `state_hash` consistent via Zobrist deltas. Returns the new hash.
    pub fn apply_advance(&mut self, p1: Direction, p2: Direction) -> u64 {
        self.engine.process_turn(p1, p2);
        self.player1_last_move = p1;
        self.player2_last_move = p2;
        self.engine.state_hash()
    }

    /// Reload the engine from a `TurnState` (received via GoState or FullState).
    /// Recomputes the hash from scratch and returns it.
    pub fn load_turn_state(&mut self, ts: &TurnState) -> u64 {
        apply_turn_state(&mut self.engine, ts);
        self.player1_last_move = ts.player1_last_move;
        self.player2_last_move = ts.player2_last_move;
        self.engine.state_hash()
    }

    /// Reload from a fresh `MatchConfig` and `TurnState` (FullState recovery).
    /// Rebuilds the engine and recomputes the hash. Returns the new hash.
    pub fn load_full_state(&mut self, cfg: &MatchConfig, ts: &TurnState) -> Result<u64, String> {
        self.engine = build_engine(cfg)?;
        self.move_timeout_ms = cfg.move_timeout_ms;
        self.preprocessing_timeout_ms = cfg.preprocessing_timeout_ms;
        Ok(self.load_turn_state(ts))
    }

    // ── Perspective helpers ─────────────────────────

    fn pick<T: Copy>(&self, p1: T, p2: T) -> T {
        match self.my_player {
            Player::Player1 => p1,
            _ => p2,
        }
    }

    fn pick_opponent<T: Copy>(&self, p1: T, p2: T) -> T {
        match self.my_player {
            Player::Player1 => p2,
            _ => p1,
        }
    }

    // ── Perspective-mapped accessors ─────────────────

    /// Which player this bot controls (assigned by `Welcome`).
    pub fn my_player(&self) -> Player {
        self.my_player
    }

    pub fn my_position(&self) -> Coordinates {
        self.pick(self.player1_position(), self.player2_position())
    }

    pub fn opponent_position(&self) -> Coordinates {
        self.pick_opponent(self.player1_position(), self.player2_position())
    }

    pub fn my_score(&self) -> f32 {
        self.pick(self.player1_score(), self.player2_score())
    }

    pub fn opponent_score(&self) -> f32 {
        self.pick_opponent(self.player1_score(), self.player2_score())
    }

    pub fn my_mud_turns(&self) -> u8 {
        self.pick(self.player1_mud_turns(), self.player2_mud_turns())
    }

    pub fn opponent_mud_turns(&self) -> u8 {
        self.pick_opponent(self.player1_mud_turns(), self.player2_mud_turns())
    }

    pub fn my_last_move(&self) -> Direction {
        self.pick(self.player1_last_move, self.player2_last_move)
    }

    pub fn opponent_last_move(&self) -> Direction {
        self.pick_opponent(self.player1_last_move, self.player2_last_move)
    }

    // ── Raw (objective) accessors ────────────────────

    pub fn player1_position(&self) -> Coordinates {
        self.engine.player1.current_pos
    }

    pub fn player2_position(&self) -> Coordinates {
        self.engine.player2.current_pos
    }

    pub fn player1_score(&self) -> f32 {
        self.engine.player1.score
    }

    pub fn player2_score(&self) -> f32 {
        self.engine.player2.score
    }

    pub fn player1_mud_turns(&self) -> u8 {
        self.engine.player1.mud_timer
    }

    pub fn player2_mud_turns(&self) -> u8 {
        self.engine.player2.mud_timer
    }

    pub fn player1_last_move(&self) -> Direction {
        self.player1_last_move
    }

    pub fn player2_last_move(&self) -> Direction {
        self.player2_last_move
    }

    pub fn turn(&self) -> u16 {
        self.engine.turn
    }

    pub fn max_turns(&self) -> u16 {
        self.engine.max_turns
    }

    pub fn cheese(&self) -> Vec<Coordinates> {
        self.engine.cheese_positions()
    }

    pub fn move_timeout_ms(&self) -> u32 {
        self.move_timeout_ms
    }

    pub fn preprocessing_timeout_ms(&self) -> u32 {
        self.preprocessing_timeout_ms
    }

    pub fn state_hash(&self) -> u64 {
        self.engine.state_hash()
    }

    // ── Convenience (delegate to a borrowed Maze) ────

    pub fn width(&self) -> u8 {
        self.engine.width
    }

    pub fn height(&self) -> u8 {
        self.engine.height
    }

    fn maze(&self) -> Maze<'_> {
        Maze::new(
            &self.engine.move_table,
            &self.engine.mud,
            self.engine.width,
            self.engine.height,
        )
    }

    /// Directions from `pos` that don't hit a wall or boundary.
    /// Defaults to `my_position()` if `pos` is `None`.
    pub fn effective_moves(&self, pos: Option<Coordinates>) -> Vec<Direction> {
        self.maze()
            .effective_moves(pos.unwrap_or_else(|| self.my_position()))
    }

    /// Cost (in turns) of moving in `dir` from `pos`.
    /// Defaults to `my_position()` if `pos` is `None`.
    pub fn move_cost(&self, dir: Direction, pos: Option<Coordinates>) -> Option<u8> {
        self.maze()
            .move_cost(pos.unwrap_or_else(|| self.my_position()), dir)
    }

    /// Shortest path with full direction sequence between two cells.
    pub fn shortest_path(&self, from: Coordinates, to: Coordinates) -> Option<FullPathResult> {
        pyrat_engine_interface::shortest_path_full(from, to, &self.maze())
    }

    /// Nearest cheese from `pos`. Returns the full path to the closest cheese.
    /// Defaults to `my_position()` if `pos` is `None`.
    ///
    /// When multiple cheeses tie at the minimum distance, returns the first one
    /// in the cheese list. Use [`nearest_cheeses`](Self::nearest_cheeses) to get
    /// all tied results.
    pub fn nearest_cheese(&self, pos: Option<Coordinates>) -> Option<FullPathResult> {
        self.nearest_cheeses(pos).into_iter().next()
    }

    /// All cheeses tied at the minimum distance from `pos`, each with a full
    /// direction sequence. Defaults to `my_position()` if `pos` is `None`.
    pub fn nearest_cheeses(&self, pos: Option<Coordinates>) -> Vec<FullPathResult> {
        let from = pos.unwrap_or_else(|| self.my_position());
        let cheese = self.engine.cheese_positions();
        pyrat_engine_interface::nearest_cheeses_full(from, &cheese, &self.maze())
    }

    /// Weighted distances from `pos` to all reachable cells.
    /// Defaults to `my_position()` if `pos` is `None`.
    pub fn distances_from(&self, pos: Option<Coordinates>) -> HashMap<Coordinates, u32> {
        pyrat_engine_interface::distances_from(
            pos.unwrap_or_else(|| self.my_position()),
            &self.maze(),
        )
    }

    /// Clone the game into a mutable simulation state.
    pub fn to_sim(&self) -> GameSim {
        self.engine.clone()
    }
}

fn build_engine(cfg: &MatchConfig) -> Result<pyrat::GameState, String> {
    let mut wall_map: HashMap<Coordinates, Vec<Coordinates>> = HashMap::new();
    for (a, b) in &cfg.walls {
        wall_map.entry(*a).or_default().push(*b);
        wall_map.entry(*b).or_default().push(*a);
    }

    let mut mud_map = MudMap::new();
    for m in &cfg.mud {
        mud_map.insert(m.pos1, m.pos2, m.turns);
    }

    GameBuilder::new(cfg.width, cfg.height)
        .with_max_turns(cfg.max_turns)
        .with_custom_maze(wall_map, mud_map)
        .with_custom_positions(cfg.player1_start, cfg.player2_start)
        .with_custom_cheese(cfg.cheese.clone())
        .build()
        .create(None)
        .map_err(|e| e.to_string())
}

fn apply_turn_state(engine: &mut pyrat::GameState, ts: &TurnState) {
    engine.player1.current_pos = ts.player1_position;
    engine.player1.score = ts.player1_score;
    engine.player1.mud_timer = ts.player1_mud_turns;

    engine.player2.current_pos = ts.player2_position;
    engine.player2.score = ts.player2_score;
    engine.player2.mud_timer = ts.player2_mud_turns;

    engine.turn = ts.turn;

    // Reset cheese to the TurnState's set: take any not present in `ts.cheese`.
    for pos in engine.cheese_positions() {
        if !ts.cheese.contains(&pos) {
            engine.cheese.take_cheese(pos);
        }
    }

    engine.recompute_state_hash();
}

#[cfg(test)]
mod tests {
    use super::*;
    use pyrat_protocol::TurnState;
    use pyrat_wire::TimingMode;

    fn test_config() -> MatchConfig {
        MatchConfig {
            width: 5,
            height: 5,
            max_turns: 300,
            walls: vec![],
            mud: vec![],
            cheese: vec![Coordinates::new(2, 2), Coordinates::new(4, 4)],
            player1_start: Coordinates::new(0, 0),
            player2_start: Coordinates::new(4, 4),
            controlled_players: vec![],
            timing: TimingMode::Wait,
            move_timeout_ms: 1000,
            preprocessing_timeout_ms: 5000,
        }
    }

    fn test_turn_state() -> TurnState {
        TurnState {
            turn: 5,
            player1_position: Coordinates::new(1, 1),
            player2_position: Coordinates::new(3, 3),
            player1_score: 1.0,
            player2_score: 0.5,
            player1_mud_turns: 0,
            player2_mud_turns: 2,
            cheese: vec![Coordinates::new(4, 4)],
            player1_last_move: Direction::Right,
            player2_last_move: Direction::Left,
        }
    }

    #[test]
    fn from_config_starts_at_turn_zero() {
        let cfg = test_config();
        let state = GameState::from_config(Player::Player1, &cfg).unwrap();

        assert_eq!(state.turn(), 0);
        assert_eq!(state.my_position(), Coordinates::new(0, 0));
        assert_eq!(state.opponent_position(), Coordinates::new(4, 4));
        assert_eq!(state.cheese().len(), 2);
    }

    #[test]
    fn load_turn_state_replaces_dynamic_fields() {
        let cfg = test_config();
        let mut state = GameState::from_config(Player::Player1, &cfg).unwrap();
        state.load_turn_state(&test_turn_state());

        assert_eq!(state.turn(), 5);
        assert_eq!(state.my_position(), Coordinates::new(1, 1));
        assert_eq!(state.opponent_position(), Coordinates::new(3, 3));
        assert!((state.my_score() - 1.0).abs() < f32::EPSILON);
        assert!((state.opponent_score() - 0.5).abs() < f32::EPSILON);
        assert_eq!(state.my_mud_turns(), 0);
        assert_eq!(state.opponent_mud_turns(), 2);
        assert_eq!(state.my_last_move(), Direction::Right);
        assert_eq!(state.opponent_last_move(), Direction::Left);
        assert_eq!(state.cheese().len(), 1);
    }

    #[test]
    fn perspective_player2() {
        let cfg = test_config();
        let mut state = GameState::from_config(Player::Player2, &cfg).unwrap();
        state.load_turn_state(&test_turn_state());

        assert_eq!(state.my_player(), Player::Player2);
        assert_eq!(state.my_position(), Coordinates::new(3, 3));
        assert_eq!(state.opponent_position(), Coordinates::new(1, 1));
        assert!((state.my_score() - 0.5).abs() < f32::EPSILON);
        assert!((state.opponent_score() - 1.0).abs() < f32::EPSILON);
        assert_eq!(state.my_mud_turns(), 2);
        assert_eq!(state.opponent_mud_turns(), 0);
    }

    #[test]
    fn raw_accessors() {
        let mut state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        state.load_turn_state(&test_turn_state());

        assert_eq!(state.player1_position(), Coordinates::new(1, 1));
        assert_eq!(state.player2_position(), Coordinates::new(3, 3));
        assert!((state.player1_score() - 1.0).abs() < f32::EPSILON);
        assert!((state.player2_score() - 0.5).abs() < f32::EPSILON);
    }

    #[test]
    fn effective_moves_from_corner() {
        let state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        let moves = state.effective_moves(None);
        assert!(moves.contains(&Direction::Right));
        assert!(moves.contains(&Direction::Up));
        assert!(!moves.contains(&Direction::Left));
        assert!(!moves.contains(&Direction::Down));
    }

    #[test]
    fn nearest_cheese_works() {
        let state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        let result = state.nearest_cheese(None).unwrap();
        // From (0,0), nearest cheese is (2,2) at distance 4.
        assert_eq!(result.target, Coordinates::new(2, 2));
        assert_eq!(result.cost, 4);
        assert!(!result.path.is_empty());
    }

    #[test]
    fn to_sim_make_unmake() {
        let state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        let mut sim = state.to_sim();

        let p1_before = sim.player1_position();
        let undo = sim.make_move(Direction::Right, Direction::Stay);
        assert_ne!(sim.player1_position(), p1_before);

        sim.unmake_move(undo);
        assert_eq!(sim.player1_position(), p1_before);
    }

    #[test]
    fn to_sim_reflects_loaded_state() {
        let mut state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        state.load_turn_state(&test_turn_state());

        let sim = state.to_sim();
        assert_eq!(sim.player1_position(), Coordinates::new(1, 1));
        assert_eq!(sim.player2_position(), Coordinates::new(3, 3));
        assert!((sim.player1_score() - 1.0).abs() < f32::EPSILON);
        assert_eq!(sim.turn, 5);
        assert_eq!(sim.cheese_positions().len(), 1);
    }

    #[test]
    fn to_sim_preserves_total_cheese() {
        let cfg = test_config();
        let mut state = GameState::from_config(Player::Player1, &cfg).unwrap();
        let original_cheese_count = cfg.cheese.len();

        // Mid-game: one cheese collected, p1 has score.
        state.load_turn_state(&TurnState {
            turn: 10,
            player1_position: Coordinates::new(2, 2),
            player2_position: Coordinates::new(0, 0),
            player1_score: 1.0,
            player2_score: 0.0,
            player1_mud_turns: 0,
            player2_mud_turns: 0,
            cheese: vec![Coordinates::new(4, 4)],
            player1_last_move: Direction::Stay,
            player2_last_move: Direction::Stay,
        });

        let sim = state.to_sim();
        assert_eq!(
            sim.cheese.total_cheese() as usize,
            original_cheese_count,
            "load_turn_state must preserve initial_cheese_count"
        );
        assert_eq!(sim.cheese_positions().len(), 1);
        assert!(
            !sim.check_game_over(),
            "score 1.0 doesn't exceed half of 2 total cheese"
        );
    }

    #[test]
    fn distances_from_works() {
        let state = GameState::from_config(Player::Player1, &test_config()).unwrap();
        let dists = state.distances_from(Some(Coordinates::new(0, 0)));
        assert_eq!(dists[&Coordinates::new(0, 0)], 0);
        assert_eq!(dists[&Coordinates::new(1, 0)], 1);
        assert_eq!(dists[&Coordinates::new(4, 4)], 8);
    }

    /// The hash the SDK derives from MatchConfig must match the host's hash
    /// (both engines compute the same Zobrist of the same maze + initial state).
    #[test]
    fn initial_hash_is_stable() {
        let cfg = test_config();
        let a = GameState::from_config(Player::Player1, &cfg).unwrap();
        let b = GameState::from_config(Player::Player1, &cfg).unwrap();
        assert_eq!(a.state_hash(), b.state_hash());
        assert_ne!(a.state_hash(), 0, "initial hash is non-trivial");
    }

    /// `apply_advance` must agree with `load_turn_state` for the same
    /// resulting position — i.e., process_turn's incremental Zobrist update
    /// matches a from-scratch recompute.
    #[test]
    fn apply_advance_matches_load_turn_state() {
        let cfg = MatchConfig {
            width: 5,
            height: 5,
            max_turns: 300,
            walls: vec![],
            mud: vec![],
            cheese: vec![Coordinates::new(2, 2)],
            player1_start: Coordinates::new(0, 0),
            player2_start: Coordinates::new(4, 4),
            controlled_players: vec![],
            timing: TimingMode::Wait,
            move_timeout_ms: 1000,
            preprocessing_timeout_ms: 5000,
        };

        // Path A: advance one turn via process_turn.
        let mut a = GameState::from_config(Player::Player1, &cfg).unwrap();
        let advanced = a.apply_advance(Direction::Right, Direction::Left);

        // Path B: load the equivalent TurnState directly, recompute hash.
        let mut b = GameState::from_config(Player::Player1, &cfg).unwrap();
        let loaded = b.load_turn_state(&TurnState {
            turn: 1,
            player1_position: Coordinates::new(1, 0),
            player2_position: Coordinates::new(3, 4),
            player1_score: 0.0,
            player2_score: 0.0,
            player1_mud_turns: 0,
            player2_mud_turns: 0,
            cheese: vec![Coordinates::new(2, 2)],
            player1_last_move: Direction::Right,
            player2_last_move: Direction::Left,
        });

        assert_eq!(advanced, loaded);
    }
}

File: sdk/python/pyrat_sdk/state.py
"""GameState — the single rich object passed to ``think()``.

Holds an engine-backed ``GameSim`` mirror that tracks the canonical Zobrist
hash. Each ``apply_advance`` invokes the engine's ``process_turn`` so the
state hash stays in sync turn-by-turn. ``load_turn_state`` /
``load_full_state`` rebuild from a snapshot (received via GoState or
FullState) and recompute the hash from scratch.
"""

from __future__ import annotations

import copy
from enum import IntEnum
from typing import Any, NamedTuple

import numpy as np

from pyrat_sdk._engine import GameSim, PyMaze


class Direction(IntEnum):
    UP = 0
    RIGHT = 1
    DOWN = 2
    LEFT = 3
    STAY = 4

    def apply_to(self, pos: tuple[int, int]) -> tuple[int, int]:
        """Return the neighbouring cell after moving in this direction."""
        dx, dy = _DIR_DELTAS[self.value]
        return (pos[0] + dx, pos[1] + dy)


_DIR_DELTAS = {0: (0, 1), 1: (1, 0), 2: (0, -1), 3: (-1, 0), 4: (0, 0)}


class Player(IntEnum):
    PLAYER1 = 0
    PLAYER2 = 1


class PathResult(NamedTuple):
    """Result of a pathfinding query.

    Matches Rust SDK's ``FullPathResult`` — same four fields.
    """

    target: tuple[int, int]
    path: list[Direction]
    first_moves: list[Direction]
    cost: int


class GameState:
    """SDK-facing game state.

    Built once from MatchConfig + the slot assigned by ``Welcome``. The
    canonical state advances via ``apply_advance`` (one engine step,
    incremental Zobrist) or rebuilds via ``load_turn_state`` /
    ``load_full_state`` (full snapshot, recompute Zobrist).
    """

    # Static config
    width: int
    height: int
    max_turns: int
    move_timeout_ms: int
    preprocessing_timeout_ms: int

    # Maze topology and the engine mirror
    _maze: PyMaze
    _sim: GameSim
    movement_matrix: np.ndarray

    # Cached derived data
    cheese_matrix: np.ndarray

    # Slot assigned by Welcome
    _is_player1: bool

    # Last moves (engine doesn't track them)
    _player1_last_move: int
    _player2_last_move: int

    def __init__(self, slot: int, config: dict[str, Any]) -> None:
        self._is_player1 = slot == 0
        self._load_config(config)

        # Initial sim from config: starts, zero scores, no mud, turn 0.
        self._sim = self._maze.to_sim(
            config["player1_start"],
            config["player2_start"],
            0.0,
            0.0,
            0,
            0,
            config["cheese"],
            0,
        )
        self._player1_last_move = int(Direction.STAY)
        self._player2_last_move = int(Direction.STAY)
        self.cheese_matrix = _build_cheese_matrix(
            config["cheese"], self.width, self.height
        )

    # ── Mutation paths ────────────────────────────────

    def apply_advance(self, p1: Direction | int, p2: Direction | int) -> int:
        """Apply both players' moves and return the new state hash.

        Drives the engine via ``make_move`` so Zobrist stays consistent
        without a full recompute.
        """
        p1_int, p2_int = int(p1), int(p2)
        self._sim.make_move(p1_int, p2_int)
        self._player1_last_move = p1_int
        self._player2_last_move = p2_int
        self.cheese_matrix = _build_cheese_matrix(
            self._sim.cheese_positions, self.width, self.height
        )
        return self._sim.state_hash

    def load_turn_state(self, ts: dict[str, Any]) -> int:
        """Replace the sim with a snapshot from the given TurnState dict."""
        self._sim = self._maze.to_sim(
            ts["player1_position"],
            ts["player2_position"],
            ts["player1_score"],
            ts["player2_score"],
            ts["player1_mud_turns"],
            ts["player2_mud_turns"],
            ts["cheese"],
            ts["turn"],
        )
        self._player1_last_move = ts["player1_last_move"]
        self._player2_last_move = ts["player2_last_move"]
        self.cheese_matrix = _build_cheese_matrix(
            ts["cheese"], self.width, self.height
        )
        return self._sim.state_hash

    def load_full_state(
        self, config: dict[str, Any], ts: dict[str, Any]
    ) -> int:
        """Rebuild from a fresh MatchConfig + TurnState (FullState recovery)."""
        self._load_config(config)
        return self.load_turn_state(ts)

    def _load_config(self, config: dict[str, Any]) -> None:
        self.width = config["width"]
        self.height = config["height"]
        self.max_turns = config["max_turns"]
        self.move_timeout_ms = config["move_timeout_ms"]
        self.preprocessing_timeout_ms = config["preprocessing_timeout_ms"]
        self._maze = PyMaze(
            self.width,
            self.height,
            config["walls"],
            config["mud"],
        )
        self.movement_matrix = self._maze.build_movement_matrix()

    # ── Sim-backed accessors ──────────────────────────

    @property
    def turn(self) -> int:
        """Current turn number."""
        return self._sim.turn

    @property
    def state_hash(self) -> int:
        """Canonical engine Zobrist hash for the current position."""
        return self._sim.state_hash

    @property
    def cheese(self) -> list[tuple[int, int]]:
        """Cheese positions remaining on the board."""
        return self._sim.cheese_positions

    # ── My / opponent perspective ──────────────────────

    @property
    def my_position(self) -> tuple[int, int]:
        return self._sim.player1_position if self._is_player1 else self._sim.player2_position

    @property
    def opponent_position(self) -> tuple[int, int]:
        return self._sim.player2_position if self._is_player1 else self._sim.player1_position

    @property
    def my_score(self) -> float:
        return self._sim.player1_score if self._is_player1 else self._sim.player2_score

    @property
    def opponent_score(self) -> float:
        return self._sim.player2_score if self._is_player1 else self._sim.player1_score

    @property
    def my_mud_turns(self) -> int:
        return (
            self._sim.player1_mud_turns
            if self._is_player1
            else self._sim.player2_mud_turns
        )

    @property
    def opponent_mud_turns(self) -> int:
        return (
            self._sim.player2_mud_turns
            if self._is_player1
            else self._sim.player1_mud_turns
        )

    @property
    def my_last_move(self) -> Direction:
        raw = self._player1_last_move if self._is_player1 else self._player2_last_move
        return Direction(raw)

    @property
    def opponent_last_move(self) -> Direction:
        raw = self._player2_last_move if self._is_player1 else self._player1_last_move
        return Direction(raw)

    @property
    def my_player(self) -> Player:
        return Player.PLAYER1 if self._is_player1 else Player.PLAYER2

    # ── Raw player data (for HivemindBot) ─────────────

    @property
    def player1_position(self) -> tuple[int, int]:
        return self._sim.player1_position

    @property
    def player2_position(self) -> tuple[int, int]:
        return self._sim.player2_position

    @property
    def player1_score(self) -> float:
        return self._sim.player1_score

    @property
    def player2_score(self) -> float:
        return self._sim.player2_score

    @property
    def player1_mud_turns(self) -> int:
        return self._sim.player1_mud_turns

    @property
    def player2_mud_turns(self) -> int:
        return self._sim.player2_mud_turns

    @property
    def player1_last_move(self) -> Direction:
        return Direction(self._player1_last_move)

    @property
    def player2_last_move(self) -> Direction:
        return Direction(self._player2_last_move)

    # ── Layer 4: simulation ───────────────────────────

    def to_sim(self) -> GameSim:
        """Independent mutable copy of the engine state.

        Returns a clone for tree-search use (``make_move`` / ``unmake_move``)
        that doesn't disturb the SDK's internal mirror.
        """
        return copy.copy(self._sim)

    # ── Layer 2 convenience ────────────────────────────

    def effective_moves(self, pos: tuple[int, int] | None = None) -> list[Direction]:
        """Directions that don't hit a wall from *pos* (default: my position).

        Returns a list of Direction values (UP, RIGHT, DOWN, LEFT).
        Does not include STAY.
        """
        if pos is None:
            pos = self.my_position
        x, y = pos
        return [Direction(d) for d in self._maze.effective_moves(x, y)]

    def move_cost(
        self, direction: Direction | int, pos: tuple[int, int] | None = None
    ) -> int | None:
        """Cost of moving in *direction* from *pos*.

        Returns None if there's a wall, 1 for a free passage (1 turn),
        or N for a mud passage (N turns to traverse).
        """
        if pos is None:
            pos = self.my_position
        x, y = pos
        result: int | None = self._maze.move_cost(x, y, direction)
        return result

    # ── Layer 3 convenience ────────────────────────────

    def shortest_path(
        self, start: tuple[int, int], goal: tuple[int, int]
    ) -> PathResult | None:
        """Shortest path between two cells.

        Returns a ``PathResult(target, path, first_moves, cost)`` where
        *path* is the full Direction sequence and *first_moves* lists
        every direction that starts an optimal route (there may be ties).
        Returns None if unreachable.
        """
        result = self._maze.shortest_path(start, goal)
        if result is None:
            return None
        target, path, first_moves, cost = result
        return PathResult(
            target,
            [Direction(d) for d in path],
            [Direction(d) for d in first_moves],
            cost,
        )

    def nearest_cheese(self, pos: tuple[int, int] | None = None) -> PathResult | None:
        """Nearest cheese from *pos* (default: my position).

        Returns a ``PathResult(target, path, first_moves, cost)``.
        Returns None if no cheese remains.

        When multiple cheeses tie at the minimum distance, returns the first
        one in the cheese list. Use ``nearest_cheeses`` to get all tied results.
        """
        if pos is None:
            pos = self.my_position
        result = self._maze.nearest_cheese(pos, self.cheese)
        if result is None:
            return None
        target, path, first_moves, cost = result
        return PathResult(
            target,
            [Direction(d) for d in path],
            [Direction(d) for d in first_moves],
            cost,
        )

    def nearest_cheeses(self, pos: tuple[int, int] | None = None) -> list[PathResult]:
        """All cheeses tied at the minimum distance from *pos* (default: my position).

        Each result has the full path reconstructed. Returns an empty list
        if no cheese remains.
        """
        if pos is None:
            pos = self.my_position
        results = self._maze.nearest_cheeses(pos, self.cheese)
        return [
            PathResult(
                target,
                [Direction(d) for d in path],
                [Direction(d) for d in first_moves],
                cost,
            )
            for target, path, first_moves, cost in results
        ]

    def distances_from(
        self, pos: tuple[int, int] | None = None
    ) -> dict[tuple[int, int], int]:
        """Weighted distances from *pos* (default: my position) to all reachable cells.

        Returns a dict of {(x, y): cost} where cost matches shortest_path costs
        (i.e. turns, accounting for mud weights).
        """
        if pos is None:
            pos = self.my_position
        result: dict[tuple[int, int], int] = self._maze.distances_from(pos)
        return result


def _build_cheese_matrix(
    cheese: list[tuple[int, int]], width: int, height: int
) -> np.ndarray:
    """Shape ``(width, height)``, dtype uint8.  1 where cheese exists."""
    mat = np.zeros((width, height), dtype=np.uint8)
    for x, y in cheese:
        mat[x, y] = 1
    return mat
